lifting-state-up/passing-down-event-handlers.md

```diff
@@ -0,0 +1,250 @@
+
+<!-- prettier-ignore-start -->
+```js
+  const updateStudentData = updatedStudent => {
+    const students = studentData.map(student => {
+      if (student.id === updatedStudent.id) {
+        return updatedStudent;
+      } else {
+        return student;
+      }
+    });
+
+    setStudentData(students);
+  }
+```
+<!-- prettier-ignore-end -->
+
+| <div style="min-width:300px;"> Piece of Code </div> | Notes                                                                                                                                                                |
+| --------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `const updateStudentData = `                        | We're creating a function named `updateStudentData`.                                                                                                                 |
+| `updatedStudent`                                    | This function accepts one argument: an object that holds updated student data. We expect this object to have the key `id`, at the very least.                        |
+| `const students = ;`                                | For this function, we'll create a helper array, `students` to contain the updated student data.                                                                      |
+| `studentData.map(...)`                              | We'll map over `studentData`...                                                                                                                                      |
+| `student => {...}`                                 | Inside `map`, we'll reference each item with `student`...                                                                                                            |
+| `if (student.id === updatedStudent.id) {`           | If we found the student that our `updatedStudent` is updating...                                                                                                     |
+| `return updatedStudent`                             | The item we should map to our `students` array is the updated student data!                                                                                          |
+| `else { return student; }`                          | Otherwise, we should map the unchanged `student` into `students`.                                                                                                    |
+| `setStudentData(students);`                         | Ultimately, we want to update the `studentData` on state. We use our state update function, `setStudentData`, and we update it to our newly formed `students` array. |
+
+Our `App` now defines and holds a function that can update our `studentData`, named `updateStudentData`.
+
```

Comment:
suggest a callout about why the heck we're copying the whole array, and making a very important (but not obvious) point about how React decides whether to trigger a re-render.
```suggestion

### !callout-info

## Helping React Notice That `studentData` Changed

Notice that during the update, we made a helper array that is more or less a copy of the data we already had. For every student whose ID does not match the `updatedStudent`, we reuse the student data that was previously in our state. So why couldn't we simply find the matching student data in our existing state and update that?

<br/>

Because React wouldn't notice the change! For container types, like arrays and objects, if we update an inner value (an item in an array, or a field in an object) and then call the relevant set function, React will see that the reference to the object currently in state is the same reference as the object being passed into the set function. For performance reasons, it assumes that if the same reference is passed in to a set function, it _is_ the same object. It will not look within the array or object for changes!

<br/>

So to get React to notice that a value in our array or object has changed, we need to make a new outer reference, and copy the existing values into it. Then when we call a set function, it will see the change and trigger a re-render.

### !end-callout

```